### PR TITLE
New Reward Scores and Micro Batches

### PR DESCRIPTION
**Pull request by abhijeetpraveen**
This PR introduces **gradient accumulation using micro-batching** to the `grpo_iteration` training loop as well as a new reward function.

### 1. Gradient Accumulation Support
- `grpo_iteration()` now accepts a new `accumulate: bool` flag.
- When `accumulate=True`, it returns the **scalar loss** instead of performing `optimizer.step()`, allowing the main loop to manage accumulation and stepping.

### 2. Micro-Batching in Training Loop
- `train.py` has been modified to:
  - Slice large batches into micro-batches.
  - Accumulate gradients over `grad_accum_steps` before stepping the optimizer.
  - Reset gradients only after accumulation.

### 3. When running the script
New arguments:
- `--micro-batch-size`: Prompts processed per forward/backward pass (default: `2`)
- `--grad-accum-steps`: Number of micro-batches to accumulate before optimizer step (default: `4`)

### 4. Reward Score
- **Format Reward**: `0.5` if the format is correct, `0.0` otherwise.
-  **Equation Reward**: `3.0` if the equation is correct, `0.0` otherwise.
-  **Total Reward**: `Format + Equation` (i.e., ranges from `0.0` to `3.5`).
- **Accuracy**: Remains a binary `0` or `1` indicator based on full correctness.

### CODE REVIEW

File: grpo.py [hunk 1]
```diff
@@ -111,7 +117,10 @@ def grpo_iteration(
         reference_model.to("cpu")
         policy_model.to(gpu_device)
 
+    clear_cache()
+
     policy_model.train()
+    total_loss = 0
     for i in range(mu):
         logger.info(f"Update iteration: {i+1}/{mu}")
         optimizer.zero_grad()
```
Comment (Cormac-C) [inline grpo.py:126]: This zero_grad call is going to prevent the gradients from accumulating properly, probably need to wrap this in an if statement like the .step call.

File: scripts/train.py [hunk 2]
```diff
@@ -174,72 +191,104 @@ def main():
...
-                    {
-                        "epoch": epoch + 1,
-                        "batch": batch_iter,
-                        "test_mean_reward": full_rewards.mean().item(),
-                        "test_mean_accuracy": full_accuracies.mean().item(),
-                    }
+        policy_model.train()
+        for big_batch in tqdm(train_dataloader, desc="Training"):
+            optimizer.zero_grad()
+            num_prompts = len(big_batch["prompt"])
```
Comment (Cormac-C) [inline scripts/train.py:216]: This vaiable is always just batch_size right? Seems redundant

File: scripts/train.py [hunk 3]
```diff
@@ -174,72 +191,104 @@ def main():
...
+                    reward_model=batch_compute_metrics,
+                    tokenizer=tokenizer,
+                    optimizer=optimizer,
+                    G=G,
+                    eps=eps,
+                    beta=beta,
+                    mu=mu,
+                    accumulate=True,
                 )
+                (loss / grad_accum_steps).backward()
```
Comment (Cormac-C) [inline scripts/train.py:234]: I think this is a duplicate call of `.backward()` since it is already called within `grpo_iteration`

Comment (Cormac-C) [reply scripts/train.py:234]: I'm also not sure I understand the logic behind diving the loss by `grad_accum_steps`

File: scripts/train.py [hunk 4]
```diff
@@ -174,72 +191,104 @@ def main():
...
+                    step_count += 1
+                    logger.info("Step %d completed.", step_count)
+                    logger.info("Train loss: %f", total_loss / grad_accum_steps)
+                    wandb.log({"train_loss": total_loss / grad_accum_steps, "step": step_count})
+                    total_loss = 0
+                
+                if step_count % EVALUATION_FREQUENCY == 0:
+                    logger.info("Batch %d/%d completed.", step_count, len(train_dataloader))
+                    logger.info("Evaluating model...")
+                    policy_model.eval()
```
Comment (Cormac-C) [inline scripts/train.py:250]: This is redundant with the first line of `evaluate_policy`

File: scripts/train.py [hunk 5]
```diff
@@ -174,72 +191,104 @@ def main():
...
+                    full_format_rewards, full_equation_rewards, full_total_rewards, full_accuracies = [], [], [], []
+                    with torch.no_grad():
```
Comment (Cormac-C) [inline scripts/train.py:252]: Same comment as above, `torch.no_grad` is used within `evaluate_policy` is there a reason we're doubling it?

**Review by Cormac-C — CHANGES_REQUESTED** [review]
Hi Abhijeet, thanks for the work. I've left some comments, I think there are a number of redundant calls to various parts of the optimization process that will likely cause problems. These kind of stem from a broader issue that I don't think having the optimization happening in two different parts of the scripts depending on a flag is the best pattern, it'd be nice to be able to unify this logic.

Have you run this code locally at all? I haven't tried that yet but it'd be a good step before merging to also weed out any runtime issues.
